execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,495 @@
+use casper_types::{Key, StoredValue};
+
+use crate::{
+    shared::{additive_map::AdditiveMap, newtypes::CorrelationId, transform::Transform},
+    storage::{
+        error,
+        global_state::{CommitError, CommitProvider, StateProvider, StateReader},
+        store::Store,
+        transaction_source::{lmdb::LmdbEnvironment, Transaction, TransactionSource},
+        trie::{merkle_proof::TrieMerkleProof, Trie},
+        trie_store::{
+            lmdb::LmdbTrieStore,
+            operations::{
+                keys_with_prefix, missing_trie_keys, put_trie, read, read_with_proof, ReadResult,
+            },
+        },
+    },
+};
+
+type SharedCache = Arc<RwLock<Cache>>;
+
+struct Cache {
+    stored_values: HashMap<Key, StoredValue>,
+}
+
+impl Cache {
+    fn new() -> Self {
+        Cache {
+            stored_values: HashMap::new(),
+        }
+    }
+
+    fn insert(&mut self, key: Key, value: StoredValue) {
+        self.stored_values.insert(key, value);
+    }
+
+    fn get(&self, key: &Key) -> Option<&StoredValue> {
+        self.stored_values.get(key)
+    }
+
+    fn into_inner(self) -> HashMap<Key, StoredValue> {
+        self.stored_values
+    }
+}
+
+/// Global state implemented against LMDB as a backing data store.
+pub struct ScratchGlobalState {
+    /// Underlying, cached stored values.
+    cache: SharedCache,
+    /// Environment for LMDB.
+    pub(crate) environment: Arc<LmdbEnvironment>,
+    /// Trie store held within LMDB.
+    pub(crate) trie_store: Arc<LmdbTrieStore>,
+    // TODO: make this a lazy-static
+    /// Empty root hash used for a new trie.
+    pub(crate) empty_root_hash: Digest,
+}
+
+/// Represents a "view" of global state at a particular root hash.
+pub struct ScratchGlobalStateView {
+    cache: SharedCache,
+    /// Environment for LMDB.
+    pub(crate) environment: Arc<LmdbEnvironment>,
+    /// Trie store held within LMDB.
+    pub(crate) trie_store: Arc<LmdbTrieStore>,
+    /// Root hash of this "view".
+    pub(crate) root_hash: Digest,
+}
+
+impl ScratchGlobalState {
+    /// Creates a state from an existing environment, store, and root_hash.
+    /// Intended to be used for testing.
+    pub fn new(
+        environment: Arc<LmdbEnvironment>,
+        trie_store: Arc<LmdbTrieStore>,
+        empty_root_hash: Digest,
+    ) -> Self {
+        ScratchGlobalState {
+            cache: Arc::new(RwLock::new(Cache::new())),
+            environment,
+            trie_store,
+            empty_root_hash,
+        }
+    }
+
+    /// Consume self and return inner cache.
+    pub fn into_inner(self) -> HashMap<Key, StoredValue> {
+        let cache = mem::replace(&mut *self.cache.write().unwrap(), Cache::new());
+        cache.into_inner()
+    }
+}
+
+impl StateReader<Key, StoredValue> for ScratchGlobalStateView {
+    type Error = error::Error;
+
+    fn read(
+        &self,
+        correlation_id: CorrelationId,
+        key: &Key,
+    ) -> Result<Option<StoredValue>, Self::Error> {
```

Comment:
We need a unit test for this method - i.e. when reading a not-yet-cached key from `ScratchGlobalStateView` it should read from the LMDB and update the cache. Since in practice we can't mock the underlying `trie_store` at that `root_hash`, it should be OK to simply verify that reading from `ScratchGlobalStateView` returns correct values.

Reply:
We do test this, in `transfer_cached`.

Reply:
`transfer_cached` is a higher level test. We could test it as a unit test here as well. The other tests are very similar and almost do that already.

Reply:
This is now covered.